src/client.js

```diff
@@ -137,23 +104,43 @@ class Client extends events.EventEmitter {
     }
 
     if (this.running) {
-      const newConn = new Connection(this);
-
-      // don't actually start connecting until at lesat minReconnectionInterval has passed
-      const earliestConnectionTime = this.lastConnectionTime + this._minReconnectionInterval;
-      const now = new Date().getTime();
-      setTimeout(() => {
-        this.lastConnectionTime = new Date().getTime();
-        newConn.connect();
-      }, now < earliestConnectionTime ? earliestConnectionTime - now : 0);
-
-      newConn.once('connected', () => {
-        this.emit('connected', newConn);
-      });
-      newConn.once('finished', () => {
-        this.connections = this.connections.filter(conn => conn !== newConn);
-      });
-      this.connections.unshift(newConn);
+      // If recycle is called due to recycleAt then refresh the normal recycle interval
+      if (this.recycleAt && this.recycleAt - new Date() === 0) {
+        clearInterval(this._interval);
+        this._interval = setInterval(
+          () => this.recycle(),
+          this._recycleInterval);
+      }
+
+      try {
+        const {connectionString, recycleAt} = await this._fetchCredentials();
+        if (recycleAt) {
+          this.recycleAt = recycleAt;
+          this._setRecycleAt();
+        }
+        const newConn = new Connection(connectionString, this._retirementDelay);
+
+        // don't actually start connecting until at least minReconnectionInterval has passed
+        const earliestConnectionTime = this.lastConnectionTime + this._minReconnectionInterval;
+        const now = new Date().getTime();
+        setTimeout(() => {
+          this.lastConnectionTime = new Date().getTime();
+          newConn.connect();
+        }, now < earliestConnectionTime ? earliestConnectionTime - now : 0);
+
+        newConn.once('connected', () => {
+          this.emit('connected', newConn);
+        });
+        newConn.once('finished', () => {
+          this.connections = this.connections.filter(conn => conn !== newConn);
+        });
+        newConn.once('failed', () => {
+          this.recycle();
+        });
+        this.connections.unshift(newConn);
+      } catch (err) {
+        return err;
+      }
```

Comment:
Returning an error (instead of throwing it) is pretty unusual.  I guess this is to catch errors from fetchCredentials?  Let's handle that directly..